CppCoreGuidelines.md

```diff
@@ -10005,19 +10004,33 @@ This makes the function's ownership sharing explicit.
     class WidgetUser
     {
     public:
-        // WidgetUser will share ownership of the widget
-        explicit WidgetUser(std::shared_ptr<widget> w) noexcept:
+        // WidgetUser will share ownership of the widget, this is a "sink" function
+        // following F.18
+        explicit WidgetUser(std::shared_ptr<widget>&& w) noexcept:
```

Comment:
Also have pass by const ref here?  Interfaces that only accept rvalue ref's can be inconvient to use.  It is possible that someone wanting to call a constructor like this would have an lvalue.

Reply:
F.18 mentions that sink functions like this one should take arguments by rvalue reference.  I'm not sure whether users are required to provide another overload with const ref, I believe it is allowed but not required.  I tried to clarify this on [my other PR](https://github.com/isocpp/CppCoreGuidelines/pull/1412/files#r282968035) but that has not been landed yet.

Reply:
A sink function is requiring the caller to get an rvalue which can put a burden on the caller.  The caller may want to keep its own copy of the shared pointer while 'sharing' it with the callee.

Reply:
I don’t mind adding an overload but if this is a general requirement we should also modify F.18 to make that clear.  As written F.18 doesn’t seem like it requires authors to add a const ref overload.

Reply:
That discussion is pretty much the motivation for recommending pass-by-value by default for shared_ptr.  If the caller has an r- value you have a move,  if not you get a copy anyway.

Reply:
How do you suggest we resolve the incompatibility of r.34 with f.16-f.18?  I understand your points and I agree but I’m struggling to synthesize them into concrete changes to this pr.

Reply:
True.  F.16-18 (F.18 specifically) do not seem to cover this kind of parameter passing.